poly1305/src/avx2.rs

```diff
@@ -0,0 +1,1305 @@
+        loop {
+            p = add4x130(
+                red4x130(mul4x130R(p, state.r4, state.r45)),
+                align4x128(load4x128(ip)),
+            );
+            in_len -= 64;
+            ip = ip.add(16);
+            if in_len < 64 {
+                break;
+            }
+        }
+    }
+}
+
+#[inline]
+#[target_feature(enable = "avx2")]
+pub unsafe fn finish(state: &mut Poly1305, output: &mut [u8]) {
+    let mut p_init: u32 = state.p_init;
+    let mut p = _mm256_set_epi64x(0, 0, 0, 0);
+    let mut buf_len = state.leftover;
+    let mut idx = 0;
+
+    if buf_len >= 32 {
+        let c = vec256x2 {
+            v0: _mm256_and_si256(
+                _mm256_or_si256(
+                    _mm256_sllv_epi32(
+                        _mm256_or_si256(
+                            _mm256_castsi128_si256(state.buffer.va()[0]),
+                            _mm256_set_epi64x(0, 1, 0, 0),
+                        ),
+                        _mm256_set_epi32(32, 32, 32, 24, 18, 12, 6, 0),
+                    ),
+                    _mm256_permutevar8x32_epi32(
+                        _mm256_srlv_epi32(
+                            _mm256_or_si256(
+                                _mm256_castsi128_si256(state.buffer.va()[0]),
+                                _mm256_set_epi64x(0, 1, 0, 0),
+                            ),
+                            _mm256_set_epi32(32, 32, 32, 2, 8, 14, 20, 26),
+                        ),
+                        _mm256_set_epi32(6, 5, 4, 3, 2, 1, 0, 7),
+                    ),
+                ),
+                _mm256_set_epi32(
+                    0, 0, 0, 0x3ffffff, 0x3ffffff, 0x3ffffff, 0x3ffffff, 0x3ffffff,
+                ),
+            ),
+            v1: _mm256_and_si256(
+                _mm256_or_si256(
+                    _mm256_sllv_epi32(
+                        _mm256_or_si256(
+                            _mm256_castsi128_si256(state.buffer.va()[1]),
+                            _mm256_set_epi64x(0, 1, 0, 0),
+                        ),
+                        _mm256_set_epi32(32, 32, 32, 24, 18, 12, 6, 0),
+                    ),
+                    _mm256_permutevar8x32_epi32(
+                        _mm256_srlv_epi32(
+                            _mm256_or_si256(
+                                _mm256_castsi128_si256(state.buffer.va()[1]),
+                                _mm256_set_epi64x(0, 1, 0, 0),
+                            ),
+                            _mm256_set_epi32(32, 32, 32, 2, 8, 14, 20, 26),
+                        ),
+                        _mm256_set_epi32(6, 5, 4, 3, 2, 1, 0, 7),
+                    ),
+                ),
+                _mm256_set_epi32(
+                    0, 0, 0, 0x3ffffff, 0x3ffffff, 0x3ffffff, 0x3ffffff, 0x3ffffff,
+                ),
+            ),
+        };
+
+        let p = red5x64(mul2x130(c, state.r1, state.r2, state.r15, state.r25));
+        idx += 2;
+        buf_len -= 32;
+        p_init += 1;
+    }
```

Comment:
It looks like an optimization which only wants to implement the current `UniversalHash` trait (which precludes optimizations for outputs larger than the UHF itself) could omit this entire codeblock and also any functions which are only invoked for this "wide" optimization.